### PR TITLE
[WIP] Proof of concept: Forward tracing in PyTorch

### PR DESCRIPTION
**Pull request by ezyang**
This commit series turns on forward trace construction in PyTorch. These forward traces are automatically saved as you execute your Torch code, and then can be rerun in the future with different input values to do a different computation. This is a PROOF OF CONCEPT because you don't actually want tracing to be turned on all the time.

Some primary characteristics of the implementation:

* The first four commits are just some refactoring commits, I'll submit them separately.
* The basic IR model is a multigraph, where each Node represents a computation, taking in some number of tensors, and producing some number of tensors. An Output represents a *particular* output of a function. Unlike the current backwards representation, the pointers in this representation go "the opposite way." Input nodes are identified by pointer equality; every Variable is associated with a unique input node which lets you get a handle into the parameters of the graph.
* The IR is exposed at the Python level using proxy objects; this means that if you don't access the trace from Python, Python objects are never materialized. Note that PyNode operators themselves contain a pointer to the Python class which implements their forward pass, but this cannot cause cycles to arise.
* The IR today uses shared pointers. When we add "regions" to trace over, it should become possible to region allocate the IR and then remove the uses of the shared pointers, to make things faster.
* The execution engine is a simple but inefficient memoized, recursive interpreter. In principle we can parallelize it in the same way today's engine is done, but I have not done this yet. The interpreter basically calls into THPFunction_apply, doing the dumbest possible thing (constructing Python objects to pass in, and then destructing Python objects on the way out.)

Where it's going from here:

* Turn tracing on/off by the user
* Add frontend support for "dumb" just-in-time compilation of operators, with no correctness checking
* Implement a peephole rewriter on the resulting trace, allowing us to apply fusion
* Port backward traces to use the new IR format
* Bikeshed the names Output/Node
* Better trace pretty-printing capabilities and a textual language design
* Figure out how to get speed increases even when operators are implemented in Python

CC @lantiga @apaszke @zdevito

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -54,7 +54,7 @@ def _single_compile(obj):
         src, ext = build[obj]
         self._compile(obj, src, ext, cc_args, extra_postargs, pp_opts)
     num_jobs = multiprocessing.cpu_count()
-    multiprocessing.pool.ThreadPool(num_jobs).map(_single_compile, objects)
+    list(multiprocessing.pool.ThreadPool(num_jobs).imap(_single_compile, objects))
```
Comment (apaszke) [inline setup.py]: Why did you need to change that?

File: torch/autograd/ir.py [hunk 2]
```diff
@@ -0,0 +1,4 @@
+import torch._C as _C
+
+class Node(_C._NodeBase):
```
Comment (apaszke) [inline torch/autograd/ir.py]: So I think `Node` is not going to be a good name for the thing. When talking about autograd, we often say 'node' when we think about Functions, and this is going to make it quite ambiguous. Perhaps `IRNode` would be better?

File: torch/csrc/autograd/ir.h [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+#include "torch/csrc/utils/object_ptr.h"
+
+namespace torch { namespace autograd {
+
+struct Output;
+struct Node;
+using output_list = std::vector<Output>;
+
+struct Output {
+  std::shared_ptr<Node> node;
```
Comment (apaszke) [inline torch/csrc/autograd/ir.h]: It seems that you have a reference cycle here. You have `Node` holding its `Output`s in a vector, and each output holds an owning pointer to the node.

File: torch/csrc/autograd/ir.h [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+struct PyNode : public Node {
+  PyNode(PyObject* pyobj, std::vector<Output> inputs)
+    : Node(inputs)
+    , pyobj(pyobj)
+    {}
+
+  virtual std::string name() const override;
+  THPObjectPtr pyobj;
+};
```
Comment (apaszke) [inline torch/csrc/autograd/ir.h]: So the idea is that it holds Functions right? Why do you hold them as Python objects and not as `torch::autograd::Function` shared_ptrs?

File: torch/csrc/autograd/python_engine.cpp [hunk 5]
```diff
@@ -225,6 +226,55 @@ PyObject *THPEngine_run_backward(THPEngine *self, PyObject *args, PyObject *kwar
   END_HANDLE_TH_ERRORS
 }
 
+PyObject *THPEngine_run_forward(THPEngine *self, PyObject *args, PyObject *kwargs)
+{
+  HANDLE_TH_ERRORS;
+  PyObject* input_objs;
+  PyObject* output_objs;
+  const char *accepted_kwargs[] = {"inputs", "outputs", NULL};
```
Comment (apaszke) [inline torch/csrc/autograd/python_engine.cpp]: So is it this function like "re-execute part of the graph between inputs and outputs"?

File: torch/csrc/autograd/python_function.cpp [hunk 6]
```diff
@@ -681,7 +682,7 @@ PyObject *THPFunction_do_forward(THPFunction *self, PyObject *_inputs)
   THPObjectPtr raw_output(PyObject_CallObject(forward_fn, unpacked_input.tensor_input));
   if (!raw_output) return NULL;
 
-  return process_outputs(self, unpacked_input, std::move(raw_output), is_volatile);
+  return process_outputs(self, unpacked_input, std::move(raw_output), std::move(nullptr), is_volatile);
```
Comment (apaszke) [inline torch/csrc/autograd/python_function.cpp]: move nullptr?

File: torch/csrc/autograd/python_function.cpp [hunk 7]
```diff
@@ -720,7 +721,20 @@ PyObject *THPFunction_apply(PyObject *cls, PyObject *_inputs)
...
+  for (auto v : unpacked_input.input_vars) {
+    // TODO: don't assume variables are always defined
+    if (v->trace_fn) {
+      inputs.emplace_back(v->trace_fn, v->output_nr);
+    } else {
+      inputs.emplace_back(v->get_input_node(), 0);
+    }
+  }
+  Py_INCREF(cls);
+  auto node = std::make_shared<PyNode>(cls, inputs);
```
Comment (apaszke) [inline torch/csrc/autograd/python_function.cpp]: Creating traces has to be optional

File: torch/csrc/autograd/python_function.h [hunk 8]
```diff
@@ -29,6 +29,10 @@ struct PyFunction : public Function {
   PyObject* obj;
 };
 
+using input_map = std::unordered_map<InputNode*, std::shared_ptr<Variable>>;
+using output_list = std::vector<Output>;
```
Comment (apaszke) [inline torch/csrc/autograd/python_function.h]: it'd be better to have a single definition for this. Otherwise they can get out of sync and be really confusing

File: torch/csrc/autograd/python_function.cpp [hunk 9]
```diff
@@ -1055,3 +1069,67 @@ std::shared_ptr<PyFunction> THPFunction_asFunction(THPFunction* self)
   Py_INCREF((PyObject*)self);
   return std::shared_ptr<PyFunction>(&self->cdata, Decref());
 }
+
+namespace torch { namespace autograd {
+
+using memo_map = std::unordered_map<Node*, variable_list>;
```
Comment (apaszke) [inline torch/csrc/autograd/python_function.cpp]: So we're going to remember all the intermediate outputs? That's not a very good idea, because we save _**a lot**_ of memory, by only keeping what we want. `Variable`s can be huge.

File: torch/csrc/autograd/python_function.cpp [hunk 10]
```diff
@@ -1055,3 +1069,67 @@ std::shared_ptr<PyFunction> THPFunction_asFunction(THPFunction* self)
...
+    auto vs = interpret_node(output.node, inputs, memo);
+    return vs.at(output.output_nr);
+}
+
+variable_list interpret_node(std::shared_ptr<Node> node, input_map& inputs, memo_map& memo) {
+    variable_list input_vars;
+    auto num_args = node->inputs.size();
+    input_vars.reserve(num_args);
+    for (auto input : node->inputs) {
+        input_vars.emplace_back(interpret_output(input, inputs, memo));
```
Comment (apaszke) [inline torch/csrc/autograd/python_function.cpp]: I don't think a recursive solution is a good idea. It might overflow the stack once the graph becomes very deep (say ResNet-1000)

File: torch/csrc/autograd/python_function.cpp [hunk 11]
```diff
@@ -1055,3 +1069,67 @@ std::shared_ptr<PyFunction> THPFunction_asFunction(THPFunction* self)
...
+        // Massage variables into form where we can THPFunction_apply it.
+        // While in principle we can avoid putting things into Python
+        // and then taking them out again, doing so seems to require an excess
+        // of faffing about to optimize a codepath that is already going to
+        // fundamentally be slow (since it calls into Python.) NOT. WORTH. IT.
+
+        PyObject* input_objs = PyTuple_New(num_args);
+        for (size_t i = 0; i < num_args; i++) {
+          PyTuple_SET_ITEM(input_objs, i, THPVariable_Wrap(input_vars.at(i)));
+        }
```
Comment (apaszke) [inline torch/csrc/autograd/python_function.cpp]: You seem to assume that functions only take in Variable arguments, but it's not really the case. New-style functions also accept regular Python objects (say a dim along which to perform the op). You might want to cache some additional info computed in `apply`. One of the objects remembers which inputs were Variables, and you can assume that remaining args are parametrizing the op

File: torch/csrc/autograd/python_function.cpp [hunk 12]
```diff
@@ -1055,3 +1069,67 @@ std::shared_ptr<PyFunction> THPFunction_asFunction(THPFunction* self)
...
+        THPObjectPtr output_objs{ THPFunction_apply(cls, input_objs) };
+        _ensure_tuple(output_objs);
+        auto num_outputs = PyTuple_GET_SIZE(output_objs.get());
+        variable_list output_vars;
+        output_vars.reserve(num_outputs);
+        for (Py_ssize_t i = 0; i < num_outputs; i++) {
```
Comment (apaszke) [inline torch/csrc/autograd/python_function.cpp]: It seems that you could use your wrapper here

File: torch/csrc/autograd/python_function.cpp [hunk 13]
```diff
@@ -1055,3 +1069,67 @@ std::shared_ptr<PyFunction> THPFunction_asFunction(THPFunction* self)
...
+          PyObject *output_obj = PyTuple_GET_ITEM(output_objs.get(), i);
+          if (!THPVariable_Check(output_obj)) {
+            throw std::logic_error("Unwrapped variable from THPFunction_apply");
```
Comment (apaszke) [inline torch/csrc/autograd/python_function.cpp]: I can't see how this error really tells what happened. Perhaps "THPFunction_apply returned a non-Variable object"?

File: torch/csrc/autograd/ir.h [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+
+struct InputNode : public Node {
+  InputNode()
+    : Node({})
+    {}
+  virtual std::string name() const override;
+};
+
+struct PyNode : public Node {
+  PyNode(PyObject* pyobj, std::vector<Output> inputs)
```
Comment (apaszke) [inline torch/csrc/autograd/ir.h]: You can accept a `THPObjectPtr&&` to emphasise that this constructor steals the pointer. Otherwise, it's worth adding a comment somewhere

File: torch/csrc/autograd/python_ir.cpp [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+    }
+  }
+  self->cdata.reset();
+  return 0;
+}
+
+static void THPNode_dealloc(THPNode* self)
+{
+  PyObject_GC_UnTrack(self);
+  THPNode_clear(self);
```
Comment (apaszke) [inline torch/csrc/autograd/python_ir.cpp]: It seems a bit weird to me that you're calling `clear` in here. It means that you can create a trace (purely in C), but if you then traverse it in Python for some reason, once the objects will start getting freed, it will also delete all the state in the trace (`fn->pyobj = nullptr`), even though the C objects will still be alive? Generally, you shouldn't manage other objects internal fields. It's cleaner to just release ownership, and let that (maybe) destruct them.

In general even if there's a ref cycle, resetting `self->cdata` will break it.

File: torch/csrc/autograd/variable.cpp [hunk 16]
```diff
@@ -59,6 +61,20 @@ auto Variable::get_grad_accumulator() -> std::shared_ptr<Function> {
...
+  }
+  std::lock_guard<std::mutex> lock(input_node_lock);
+
+  auto result = input_node.lock();
+  if (result) return result;
+
+  result = std::make_shared<InputNode>();
+  input_node = result;
+  return result;
+}
```
Comment (apaszke) [inline torch/csrc/autograd/variable.cpp]: I'm not sure why should we cache a `weak_ptr` to this node? It seems that it has no connection to `this` whatsoever, so it could just return a new one every time?

**Review by apaszke — COMMENTED** [review]
I wonder if there's really no way to reuse the Engine we have for backward. It seems that all `interpret` does sth very similar, but is much simpler (and less efficient)

File: torch/csrc/autograd/variable.h [hunk 17]
```diff
@@ -85,11 +87,14 @@ struct Variable : std::enable_shared_from_this<Variable> {
...
   std::unique_ptr<thpp::Tensor> data;
   std::shared_ptr<Function> grad_fn;
+  std::shared_ptr<Node> trace_fn;
   std::shared_ptr<Variable> grad;
   std::unique_ptr<VariableVersion> version_counter;
   std::vector<std::shared_ptr<FunctionPreHook>> hooks;
   std::weak_ptr<Function> grad_accumulator;
+  std::weak_ptr<InputNode> input_node;
   std::mutex grad_accumulator_lock;
+  std::mutex input_node_lock;
```
Comment (ezyang) [inline torch/csrc/autograd/variable.h]: input_node got axed!

File: torch/csrc/utils/python_tuples.h [hunk 18]
```diff
@@ -0,0 +1,56 @@
+#pragma once
+
+#include <Python.h>
+
+class PyTuple {
```
Comment (ezyang) [inline torch/csrc/utils/python_tuples.h]: I'm thinking I am just going to can this class.

File: setup.py (hunk 1 above)
Comment (ezyang) [inline setup.py]: I wanted to force all the builds to finish before exiting; otherwise I get spew from the background processes over my command prompt.

File: torch/autograd/ir.py (hunk 2 above)
Comment (ezyang) [inline torch/autograd/ir.py]: Node got renamed to Expr/Arg

File: torch/csrc/autograd/ir.h (hunk 3 above)
Comment (ezyang) [inline torch/csrc/autograd/ir.h]: This design got rewritten.

File: torch/csrc/autograd/ir.h (hunk 4 above)
Comment (ezyang) [inline torch/csrc/autograd/ir.h]: I think this is just to reduce the amount of refactoring we need to do. To execute PyApply nodes, I call THPFunction_do_forward or THPFunction_apply; in both cases they take PyObject of some sort (THPFunction in the former case, but hardly a Function.)

File: torch/csrc/autograd/python_engine.cpp (hunk 5 above)
Comment (ezyang) [inline torch/csrc/autograd/python_engine.cpp]: Yes. But I rewrote this so there's a new version of the function

File: torch/csrc/autograd/python_function.cpp (hunk 6 above)
Comment (ezyang) [inline torch/csrc/autograd/python_function.cpp]: This got killed

File: torch/csrc/autograd/python_function.cpp (hunk 7 above)
Comment (ezyang) [inline torch/csrc/autograd/python_function.cpp]: Now implemented in latest branch

File: torch/csrc/autograd/python_function.cpp (hunk 9 above)
Comment (ezyang) [inline torch/csrc/autograd/python_function.cpp]: This is still a bug

File: torch/csrc/autograd/python_function.cpp (hunk 10 above)
Comment (ezyang) [inline torch/csrc/autograd/python_function.cpp]: The interpreter is now better in this respect, it's been rewritten to execute a series of let statements; it's still recursive but TCO applies.

File: torch/csrc/autograd/python_function.cpp (hunk 11 above)
Comment (ezyang) [inline torch/csrc/autograd/python_function.cpp]: Yes. The rewrite handles this correctly, but @zdevito convinced me that I should store the arguments out of band, and reconstruct the correct Python argument positions on the fly.

File: torch/csrc/autograd/python_function.cpp (hunk 12 above)
Comment (ezyang) [inline torch/csrc/autograd/python_function.cpp]: Nah ;)

File: torch/csrc/autograd/python_function.h (hunk 8 above)
Comment (ezyang) [inline torch/csrc/autograd/python_function.h]: this got axed

File: torch/csrc/autograd/python_ir.cpp (hunk 15 above)
Comment (ezyang) [inline torch/csrc/autograd/python_ir.cpp]: This is still a problem

File: torch/csrc/autograd/variable.cpp (hunk 16 above)
Comment (ezyang) [inline torch/csrc/autograd/variable.cpp]: This code got axed

File: torch/csrc/autograd/python_ir.cpp (hunk 15 above)
Comment (ezyang) [inline torch/csrc/autograd/python_ir.cpp]: @apaszke So, It looks like I cargo-culted this code from THPVariable_dealloc. But I guess indeed we should not have a clear here; the clear is necessary for THPVariable because there is only at most one Python object per Variable, but that invariant doesn't hold here.

**Comment by ryanleary** [comment]
@ezyang are you still working on this?

**Comment by ezyang** [comment]
Yep. Dev is still proceeding apace in the jit branch.
